prog/analysis.go

```diff
@@ -309,3 +310,98 @@ func encodeFallbackSignal(typ, id, aux int) uint32 {
 func decodeFallbackSignal(s uint32) (typ, id, aux int) {
 	return int(s & 7), int((s >> 3) & fallbackCallMask), int(s >> 16)
 }
+
+type pair struct {
+	addr uint64
+	len  uint64
+}
+
+func detectIntersection(ranges []pair) bool {
+	sort.SliceStable(ranges, func(i, j int) bool {
+		return ranges[i].addr < ranges[j].addr
+	})
+	for i := 0; i < len(ranges)-1; i++ {
+		if ranges[i+1].addr < ranges[i].addr+ranges[i].len {
+			return true
+		}
+	}
+	return false
+}
+
+func max(a, b uint64) uint64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func filterArguments(call *Call, pointers map[Arg]uint64, p *Prog) bool {
+	var ranges []pair
+	ForeachArg(call, func(arg Arg, ctx *ArgCtx) {
+		switch a := arg.(type) {
+		case *PointerArg:
+			if ctx.Base == nil {
+				return
+			}
+			value, ok := pointers[arg]
+			addr := p.Target.PhysicalAddr(ctx.Base) + ctx.Offset
+			addr -= arg.Type().UnitOffset()
+			if ok {
+				ranges = append(ranges, pair{addr, value})
+			} else if a.Res != nil {
+				ranges = append(ranges, pair{addr, max(a.Res.Size(), 1)})
+			}
+		case *GroupArg:
+			addr := p.Target.PhysicalAddr(ctx.Base) + ctx.Offset
+			addr -= arg.Type().UnitOffset()
+			ranges = append(ranges, pair{addr, max(arg.Size(), 1)})
+		}
+
+	})
+	return detectIntersection(ranges)
+}
+
+func getPointer(path []string, fields []Field, call *Call) Arg {
+	elem := path[0]
+	var pointer Arg
+	for i, buf := range call.Args {
+		if elem != fields[i].Name {
+			continue
+		}
+		//TODO: Check on pointer for invalid cases
+		pointer = buf
+		/*buf = InnerArg(buf)
+		if buf == nil {
+			dst.Val = 0 // target is an optional pointer
+			return
+		}*/
+		break
+
+	}
+	return pointer
+}
+
+func DFetchAnalysis(p *Prog) bool {
```

Comment:
I think DFetchAnalysis isn't the best name for this function, because it doesn't reflect the function's purpose.
You do not really perform the actual double fetch analysis here (it is done by the kernel), but rather check for overlapping arguments to decide whether or not we want to enable df analysis.
I would rather call it "WantDFetchAnalysis" (and return false if we need to disable it), "WantDisableDFetchAnalysis" (and return true in the same case), "HasOverlappingArgs" etc.

Reply:
You'll also need to elaborate in the comment what this function is doing.